### PR TITLE
Moved CanExecute logic to ExecuteAsync

### PR DESCRIPTION
**Pull request by haavamoa**
## Added / Fixed

- Made sure that `ExecuteAsync` will do a `CanExecute` before execution.

## Todo List

- [ ] I have tested on an Android device.
- [ ] I have tested on an iOS device.
- [ ] I have added unit tests

>*Please add pictures / Gifs if possible*
>*Todo List can be checked by putting a `X` inside the brackets*
>*Remember to update our `wiki` after this PR is merged*

### CODE REVIEW

File: src/DIPS.Xamarin.UI/Commands/AsyncCommand.cs [hunk 1]
```diff
@@ -114,25 +118,31 @@ public bool CanExecute(object parameter)
...
                 m_onException(exception);
+                return Task.CompletedTask;
             }
+        }
+
+#pragma warning disable RECS0165
+        /// <inheritdoc />
+        public async void Execute(object parameter)
+        {
+            if (!(parameter is T value)) return;
```
Comment (Illedan) [inline src/DIPS.Xamarin.UI/Commands/AsyncCommand.cs]: parameter is T  is checked in ExecuteAsync

File: src/DIPS.Xamarin.UI/Commands/AsyncCommand.cs [hunk 2]
```diff
@@ -114,25 +118,31 @@ public bool CanExecute(object parameter)
...
+            await ExecuteAsync(value);
```
Comment (Illedan) [inline src/DIPS.Xamarin.UI/Commands/AsyncCommand.cs]: Remove the async signature and the await keyword.
All Exceptions are catched in ExecuteAsync anyway.

File: src/DIPS.Xamarin.UI/Commands/AsyncCommand.cs [hunk 3]
```diff
@@ -114,25 +118,31 @@ public bool CanExecute(object parameter)
...
             {
-                if(!(parameter is T value) || !m_canExecute(value))
+                if (!(value is T actualValue) || !m_canExecute(actualValue))
                 {
-                    return;
+                    return Task.CompletedTask;
                 }
 
-                await ExecuteAsync(value);
+                return m_execute(actualValue);
```
Comment (Illedan) [inline src/DIPS.Xamarin.UI/Commands/AsyncCommand.cs]: await and remove all return Task.CompletedTask
If we return a task, the stacktrace would be buggy. Hence we need to await all tasks.

File: src/DIPS.Xamarin.UI/Commands/AsyncCommand.cs [hunk 4]
```diff
@@ -36,29 +36,33 @@ public AsyncCommand(Func<Task> execute, Func<bool> canExecute, Action<Exception>
...
             {
-                if (!CanExecute(parameter))
+                if (!CanExecute(null))
                 {
-                    return;
+                    return Task.CompletedTask;
                 }
 
-                await ExecuteAsync();
+                return m_execute();
```
Comment (Illedan) [inline src/DIPS.Xamarin.UI/Commands/AsyncCommand.cs]: await

File: src/DIPS.Xamarin.UI/Commands/AsyncCommand.cs (hunk 1 above)
Comment (haavamoa) [reply src/DIPS.Xamarin.UI/Commands/AsyncCommand.cs]: `ExecuteAsync` takes a `T`, not a `object`. So this cast is there to safely cast it.

**Pull request merged**